### PR TITLE
[rllib] Extra Changes for Usability

### PR DESCRIPTION
**Pull request by richardliaw**
Includes items such as:

 - [x] Optimizer.make
 - [x] Tensorboard fixes (was previously dropping many metrics)
 - [x] `compute_targets`
 - [x] KerasPolicyGraph

### CODE REVIEW

File: python/ray/rllib/evaluation/policy_evaluator.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def __init__(
             policy_mapping_fn=None,
             tf_session_creator=None,
             batch_steps=100,
-            batch_mode="truncate_episodes",
+            batch_mode="complete_episodes",
```
Comment (ericl) [inline python/ray/rllib/evaluation/policy_evaluator.py]: Why this change?

File: python/ray/rllib/optimizers/policy_optimizer.py [hunk 2]
```diff
@@ -105,11 +117,41 @@ def foreach_evaluator_with_index(self, func):
...
             raise NotImplementedError(
                 "This optimizer does not support multi-agent yet.")
+
+    @classmethod
+    def make(cls, num_workers=0, num_envs_per_worker=None,
+             optimizer_config=None, remote_num_cpus=None,
+             remote_num_gpus=None, *eval_args, **eval_kwargs):
+        """Creates an Optimizer with local and remote evaluators.
+
+        Args:
```
Comment (ericl) [inline python/ray/rllib/optimizers/policy_optimizer.py]: Isn't it "Arguments"?

File: python/ray/rllib/optimizers/policy_optimizer.py [hunk 3]
```diff
@@ -105,11 +117,41 @@ def foreach_evaluator_with_index(self, func):
...
+                If set, overrides `num_envs` in kwargs
+                for PolicyEvaluator.__init__.
+            optimizer_config (dict): Config passed to the optimizer.
+            remote_num_cpus (int): CPU specification for remote evaluator.
+            remote_num_gpus (int): GPU specification for remote evaluator.
+            *eval_args: PolicyEvaluator Class positional args.
+            **eval_kwargs: PolicyEvaluator Class non-positional args.
+
+        Returns:
+            (Optimizer) Instance of `cls` with evaluators c
```
Comment (ericl) [inline python/ray/rllib/optimizers/policy_optimizer.py]: Weird line break.

File: python/ray/rllib/optimizers/policy_optimizer.py [hunk 4]
```diff
@@ -105,11 +117,60 @@ def foreach_evaluator_with_index(self, func):
...
+                accordingly.
+        """
+        optimizer_config = optimizer_config or {}
+        if num_envs_per_worker:
+            assert num_envs_per_worker > 0, "Improper num_envs_per_worker!"
+            eval_kwargs["num_envs"] = int(num_envs_per_worker)
+        if optimizer_batch_size:
+            assert optimizer_batch_size > 0
+            if num_workers > 1:
+                eval_kwargs["batch_steps"] = optimizer_batch_size / num_workers
```
Comment (ericl) [inline python/ray/rllib/optimizers/policy_optimizer.py]: // for int division

File: python/ray/tune/log_sync.py [hunk 5]
```diff
@@ -66,14 +66,17 @@ class _LogSyncer(object):
...
+    def __init__(self, local_dir, remote_dir=None, verbose=True):
         self.local_dir = local_dir
         self.remote_dir = remote_dir
         self.last_sync_time = 0
         self.sync_process = None
         self.local_ip = ray.services.get_node_ip_address()
         self.worker_ip = None
-        print("Created LogSyncer for {} -> {}".format(local_dir, remote_dir))
+        self.verbose = verbose
+        if self.verbose:
```
Comment (ericl) [inline python/ray/tune/log_sync.py]: Let's not use ad-hoc log flags, either leave it as is, or switch to using logging.

File: python/ray/tune/logger.py [hunk 6]
```diff
@@ -109,7 +113,7 @@ def to_tf_values(result, path):
     values = []
     for attr, value in result.items():
         if value is not None:
-            if type(value) in [int, float]:
+            if type(value) in [int, float, np.float32, np.float64, np.int32]:
```
Comment (ericl) [inline python/ray/tune/logger.py]: Nice.

File: python/ray/rllib/evaluation/keras_policy_graph.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+from keras import backend as K
+import numpy as np
+from ray.rllib.evaluation.policy_graph import PolicyGraph
+
+
+def _sample(probs):
+    return [np.random.choice(len(pr), p=pr) for pr in probs]
+
+
+class KerasPolicyGraph(PolicyGraph):
```
Comment (ericl) [inline python/ray/rllib/evaluation/keras_policy_graph.py]: Should probably add a disclaimer that this is half-baked compared to the TF / Torch graphs.

File: python/ray/rllib/evaluation/__init__.py [hunk 8]
```diff
@@ -3,15 +3,19 @@
...
 from ray.rllib.evaluation.sampler import SyncSampler, AsyncSampler
-from ray.rllib.evaluation.postprocessing import compute_advantages
+from ray.rllib.evaluation.postprocessing import (
+    compute_advantages, compute_targets)
 from ray.rllib.evaluation.metrics import collect_metrics
 
 __all__ = [
     "EvaluatorInterface", "PolicyEvaluator", "PolicyGraph", "TFPolicyGraph",
-    "TorchPolicyGraph", "SampleBatch", "MultiAgentBatch", "SampleBatchBuilder",
+    "TorchPolicyGraph", "KerasPolicyGraph", "SampleBatch", "MultiAgentBatch",
```
Comment (ericl) [inline python/ray/rllib/evaluation/__init__.py]: Let's leave Keras out of this so it doesn't show up in docs.

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/rllib/evaluation/keras_policy_graph.py (hunk 7 above)
Comment (richardliaw) [reply python/ray/rllib/evaluation/keras_policy_graph.py]: Done.

**Review by ericl — APPROVED** [review]

**Comment by richardliaw** [comment]
Lint passes, 3.6 passes, 2.7 failing due to log overflow, and Jenkins tests failing due to `multi_node_tests`.
